runtime/src/main/kotlin/kotlin/collections/ArrayList.kt

```diff
@@ -309,7 +309,8 @@ actual class ArrayList<E> private constructor(
         }
 
         override fun set(element: E) {
-            list.checkIndex(lastIndex)
+            list.checkIsMutable()
+            check(lastIndex != -1) { "Call next() or previous() before replacing element from the iterator." }
```

Comment:
Is it possible to get `IOOBE` even after this check?

Reply:
`lastIndex` is changed to non `-1` only in `previous()` and `next()` operations. Hence if the list is not modified bypassing this iterator, this check should be enough.

Reply:
Yes, I mean the situation when the list is modified after moving the iterator. Is it possible, for example, to write the value beyond the last element of the sublist and go unnoticed?

Reply:
I believe it's not possible if the list is modified via this iterator, because iterator modifies only with `add` or `remove` operations, and they both set `lastIndex` to `-1`. 
IIRC if the list/sublist is modified bypassing its already created iterator, the iterators behaviour becomes undefined.

Reply:
Index out of bounds is also checked now